HW1/P6/P6.py

```diff
@@ -0,0 +1,48 @@
+import numpy as np
+from itertools import groupby
+
+
+def is_not_number(s):
+    try:
+        float(s)
+        return False
+    except ValueError:
+        return True
+
+def not_all_caps(s):
+	if s==s.upper():
+		return False
+	else:
+		return True
+
+def groupbyFreq(words):
+	return dict((key, len(list(group))) for key, group in groupby(sorted(words)))
+
+if __name__ == "__main__":
+	words=sc.textFile("shakespeare.txt").flatMap(lambda i:i.split()) \
+		.filter(is_not_number).filter(not_all_caps)
+	
+	#generate word group list of form (('w1','w2'),'w3')
+	w1=words.zipWithIndex().map(lambda i:(i[1],i[0]))
+	w2=w1.map(lambda i:(i[0]+1,i[1]))
+	w3=w1.map(lambda i:(i[0]+2,i[1]))
+
+	word_pairs=w3.join(w2).join(w1).values()
+	
+	phrase_freq=word_pairs.groupByKey()\
+		.map(lambda i:(i[0],groupbyFreq(i[1])))
+	phrase_list=word_pairs.groupByKey() \
+		.map(lambda i:(i[0],list(i[1]))).cache() #we'll be using phrase_list a lot, 
+													#so cache it
+
+#Generate sentences
+		sentences=''
+	for x in range(10):
+		pair1=phrase_list.takeSample(1,1)
+		sen=list(pair1[0][0])
+		sen_length=20
+		for i in range(2,sen_length):
+			sen.append(random.choice(phrase_list.lookup((sen[i-2],sen[i-1]))[0]))
+		sentences=sentences+' '.join(sen)+'\n'
+	with open("P6.txt", "w") as f:
+		f.write(sentences)
```

Comment:
I'm pretty sure I do the weighting correctly, since I don't eliminate duplicates from the list of word3's. It seemed to me like having the form (K,V)=((word1,word2),((word3a,n_3a),(word3b,n_3b),...)) was worse than just having the list (word3a,word3b,word3c,...). If you pick a random word3 weighted by n_3, I'm pretty sure that's the same as just having word3 listed n_3 times in the list and picking a random sample with no weight. The only difference is I didn't ever have to count the number of word3's there were, so it saved one step. I had thought I had mentioned that in my notes for this problem, but apparently not! Let me know if that doesn't seem right.